src/nlr/Layer.cpp

```diff
@@ -1481,4 +1501,142 @@ double Layer::getEliminatedNeuronValue( unsigned neuron ) const
     return _eliminatedNeurons[neuron];
 }
 
+void Layer::reduceIndexFromAllMaps( unsigned startIndex )
+{
+    // Adjsut the source layers
+    Map<unsigned, unsigned> copyOfSources = _sourceLayers;
+    _sourceLayers.clear();
+    for ( const auto &pair : copyOfSources )
+        _sourceLayers[pair.first >= startIndex ? pair.first - 1 : pair.first] = pair.second;
+
+    // Adjust all weight maps
+    adjustWeightMapIndexing( _layerToWeights, startIndex );
+    adjustWeightMapIndexing( _layerToPositiveWeights, startIndex );
+    adjustWeightMapIndexing( _layerToNegativeWeights, startIndex );
+
+    // Adjust the neuron activations
+    for ( auto &neuronToSources : _neuronToActivationSources )
+    {
+        for ( auto &source : neuronToSources.second )
+        {
+            if ( source._layer >= startIndex )
+                --source._layer;
+        }
+    }
+}
+
+void Layer::adjustWeightMapIndexing( Map<unsigned, double *> &map, unsigned startIndex )
+{
+    Map<unsigned, double *> copyOfWeights = map;
+    map.clear();
+    for ( const auto &pair : copyOfWeights )
+        map[pair.first >= startIndex ? pair.first - 1 : pair.first] = pair.second;
+}
+
+void Layer::reduceIndexAfterMerge (unsigned startIndex)
+{
+    if ( _layerIndex >= startIndex )
+    {
+        _layerIndex--;
+    }
+}
+
+bool Layer::operator==( const Layer & layer ) const
+{
+    if ( _layerIndex != layer._layerIndex )
+    {
+        return false;
+    }
+
+    if ( _type != layer._type )
+    {
+        return false;
+    }
+
+    if ( _size != layer._size )
+    {
+        return false;
+    }
+
+    if ( _inputLayerSize != layer._inputLayerSize )
+    {
+        return false;
+    }
+
+    if ( ( _bias && !layer._bias ) || ( !_bias && layer._bias ) )
+    {
+        return false;
+    }
+
+    if (_bias && layer._bias )
+    {
+        if ( std::memcmp(_bias, layer._bias, _size ) != 0 )
+        {
+            return false;
+        }
+    }
+
+    if ( !compareWights( _layerToWeights, layer._layerToWeights ) )
+    {
+        return false;
+    }
+
+    if (!compareWights( _layerToPositiveWeights, layer._layerToPositiveWeights ) )
+    {
+        return false;
+    }
+
+    if ( !compareWights( _layerToNegativeWeights, layer._layerToNegativeWeights ) )
+    {
+        return false;
+    }
+
+    if ( _sourceLayers != layer._sourceLayers )
+    {
+        return false;
+    }
+
+    return true;
+}
+
+bool Layer::compareWights( const Map<unsigned, double*> &map, const Map<unsigned, double*> &mapOfOtherLayer ) const
+{
+    if ( map.size() != mapOfOtherLayer.size() )
+    {
+        return false;
+    }
+
+    for ( const auto &pair : map )
+    {
+        unsigned key = pair.first;
+        double* value = pair.second;
+
+        if ( !mapOfOtherLayer.exists( key ) )
+        {
+            return false;
+        }
+
+        if ( sizeof( *value ) / sizeof( double ) != sizeof( *mapOfOtherLayer[key] ) / sizeof( double ) )
```

Comment:
Not sure what you were trying to check. I think sizeof( *value ) will just be sizeof( double ), so always 1.

Reply:
Actually it's the number of bytes in the array. And in order to have the number of elements I need to divide it byte the size of a single double

Reply:
I need the number of elements in the array. so you suggest changing it?

Reply:
OK I thought about it and we can remove these lines completely. I'll explain:
The size of each double array is the size of the current layer multiplied by the size of the input layer. BUT - we already compared (a few lines before) the sizes of the layers, and once we compare keys (input layer indices before these lines) we know the same input layer.